### PR TITLE
MM-45022 : Gekidou - Refactor code to use custom nav hooks

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
A bit of refactoring to bring about the custom hooks to most of the components.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45022

#### Device Information
This PR was tested on: 

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: app/screens/apps_form/apps_form_component.tsx [hunk 1]
```diff
@@ -125,27 +126,15 @@ function AppsFormComponent({
...
-            unsubscribe.remove();
-        };
+    const onHandleSubmit = useCallback(() => {
+        if (!submitting) {
+            handleSubmit();
+        }
     }, [serverUrl, componentId, submitting]);
 
+    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, close, [close]);
+    useNavButtonPressed(SUBMIT_BUTTON_ID, componentId, onHandleSubmit, [onHandleSubmit]);
```
Comment (enahum) [inline app/screens/apps_form/apps_form_component.tsx:136]: not a fan, this subscribes multiple listeners when one is more than enough but this makes the code cleaner for sure and I really doubt one more listener makes the difference to keep this out

File: app/screens/browse_channels/browse_channels.tsx [hunk 2]
```diff
@@ -162,27 +163,16 @@ export default function BrowseChannels(props: Props) {
...
-            unsubscribe.remove();
-        };
+    const handleCreate = useCallback(() => {
+        const screen = Screens.CREATE_OR_EDIT_CHANNEL;
+        const title = intl.formatMessage({id: 'mobile.create_channel.title', defaultMessage: 'New channel'});
+        goToScreen(screen, title);
     }, [intl.locale]);
 
+    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, close, [close]);
+
```
Comment (enahum) [inline app/screens/browse_channels/browse_channels.tsx:173]:
```suggestion
```

File: app/screens/channel_info/channel_info.tsx [hunk 3]
```diff
@@ -44,20 +44,12 @@ const ChannelInfo = ({channelId, closeButtonId, componentId, type}: Props) => {
...
-                if (buttonId === closeButtonId) {
-                    dismissModal({componentId});
-                }
-            },
-        }, componentId);
-
-        return () => {
-            update.remove();
-        };
+    const onPressed = useCallback(() => {
```
Comment (enahum) [inline app/screens/channel_info/channel_info.tsx:47]: I see no reason to memoize this

File: app/screens/create_direct_message/create_direct_message.tsx [hunk 4]
```diff
@@ -303,18 +299,9 @@ export default function CreateDirectMessage({
...
-            } else if (buttonId === CLOSE_BUTTON) {
-                close();
-            }
-        }}, componentId);
-        return () => {
-            unsubscribe.remove();
-        };
-    }, [startConversation]);
+    useNavButtonPressed(START_BUTTON, componentId, startConversation, [startConversation]);
+
```
Comment (enahum) [inline app/screens/create_direct_message/create_direct_message.tsx:303]:
```suggestion
```

File: app/screens/create_or_edit_channel/create_or_edit_channel.tsx [hunk 5]
```diff
@@ -220,27 +221,15 @@ const CreateOrEditChannel = ({
...
-    }, [onCreateChannel, onUpdateChannel, isModal]);
+    const handleClose = useCallback(() => {
+        close(componentId, isModal);
+    }, [isModal]);
+
+    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, handleClose, [handleClose]);
+
+    useNavButtonPressed(CREATE_BUTTON_ID, componentId, onCreateChannel, [onCreateChannel]);
+
+    useNavButtonPressed(EDIT_BUTTON_ID, componentId, onUpdateChannel, [onUpdateChannel]);
```
Comment (enahum) [inline app/screens/create_or_edit_channel/create_or_edit_channel.tsx:232]:
```suggestion
    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, handleClose, [handleClose]);
    useNavButtonPressed(CREATE_BUTTON_ID, componentId, onCreateChannel, [onCreateChannel]);
    useNavButtonPressed(EDIT_BUTTON_ID, componentId, onUpdateChannel, [onUpdateChannel]);
```

File: app/screens/edit_profile/edit_profile.tsx [hunk 6]
```diff
@@ -217,6 +186,9 @@ const EditProfile = ({
         scrollViewRef.current?.scrollToPosition(0, 0, true);
     }, [enableSaveButton]);
 
+    useNavButtonPressed(UPDATE_BUTTON_ID, componentId, submitUser, [userInfo]);
+    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, close, []);
```
Comment (enahum) [inline app/screens/edit_profile/edit_profile.tsx:190]: can we group this with the other hooks?

File: app/screens/emoji_picker/index.tsx [hunk 7]
```diff
@@ -14,37 +14,29 @@ type Props = {
...
-                }
-            },
-        }, componentId);
-
-        return () => {
-            unsubscribe.remove();
-        };
     }, []);
 
     const close = useCallback(() => {
```
Comment (enahum) [inline app/screens/emoji_picker/index.tsx:33]: I know is not part of the PR, but looking at this function I don't think it needs memoization, also dismissModal should include the componentId to be more certain

File: app/screens/post_options/post_options.tsx [hunk 8]
```diff
@@ -45,23 +46,12 @@ const PostOptions = ({
...
-                        break;
-                    }
-                }
-            },
-        }, componentId);
-
-        return () => {
-            unsubscribe.remove();
-        };
+    const close = useCallback(() => {
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:49]: no need to memoize

File: app/screens/settings/settings.tsx [hunk 9]
```diff
@@ -93,33 +93,9 @@ const Settings = ({componentId, showHelp, siteName}: SettingsProps) => {
...
-                    close();
-                }
-            },
-        }, componentId);
-
-        return () => {
-            unsubscribe.remove();
-        };
-    }, []);
+    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, close, [close]);
```
Comment (enahum) [inline app/screens/settings/settings.tsx:98]: close does not need to be memoized, and it would not be needed as a dep

File: app/screens/thread_options/thread_options.tsx [hunk 10]
```diff
@@ -57,20 +59,12 @@ const ThreadOptions = ({
...
-                if (buttonId === 'close-thread-options') {
-                    dismissModal({componentId});
-                }
-            },
-        }, componentId);
-
-        return () => {
-            unsubscribe.remove();
-        };
+    const close = useCallback(() => {
```
Comment (enahum) [inline app/screens/thread_options/thread_options.tsx:62]: no need to memoize

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/screens/channel_info/channel_info.tsx (hunk 3 above)
Comment (avinashlng1080) [reply app/screens/channel_info/channel_info.tsx:47]: corrected

File: app/screens/edit_profile/edit_profile.tsx (hunk 6 above)
Comment (avinashlng1080) [reply app/screens/edit_profile/edit_profile.tsx:190]: done :)

File: app/screens/emoji_picker/index.tsx (hunk 7 above)
Comment (avinashlng1080) [reply app/screens/emoji_picker/index.tsx:33]: done :)

File: app/screens/post_options/post_options.tsx (hunk 8 above)
Comment (avinashlng1080) [reply app/screens/post_options/post_options.tsx:49]: corrected

File: app/screens/settings/settings.tsx (hunk 9 above)
Comment (avinashlng1080) [reply app/screens/settings/settings.tsx:98]: corrected

File: app/screens/thread_options/thread_options.tsx (hunk 10 above)
Comment (avinashlng1080) [reply app/screens/thread_options/thread_options.tsx:62]: corrected

File: app/screens/apps_form/apps_form_component.tsx (hunk 1 above)
Comment (avinashlng1080) [reply app/screens/apps_form/apps_form_component.tsx:136]: Yes.

**Review by larkox — APPROVED** [review]
LGTM

**Review by enahum — APPROVED** [review]
